### PR TITLE
Add generic type Score to SolverManager, SE, CMT, CM and Indictment

### PR DESCRIPTION
**Pull request by ge0ffrey**
Merge these together:

* https://github.com/kiegroup/optaplanner/pull/928
* https://github.com/kiegroup/optaplanner-quickstarts/pull/3
* https://github.com/kiegroup/optaplanner-website/pull/279

### CODE REVIEW

**Comment by ge0ffrey** [comment]
Once this is merged, we can do a separate experiment to make InnerScoreDirectory have a Score_ type.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/AbstractBendableScore.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/AbstractBendableScore.java:37]: I'm not exactly sure of what this method does, but shouldn't the first argument rather be `Class<? extends AbstractBendableScore<?>>`?

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/AbstractScore.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/AbstractScore.java:39]: `Class<? extends AbstractScore<?>>`?
I'm thinking if we should just make an instance method, and use `Score_` instead.
(Likewise for other similar methods on this class.)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:686]: Now wouldn't it be nice if `SolutionDescriptor` also had a `Score_`? But maybe that'd just go too far.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/supply/SupplyManager.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/supply/SupplyManager.java:34]: First of all, `<Supply_ extends Supply> Supply_` should work, right?
Second, wouldn't it make sense to do `SupplyManager<Supply_>` directly?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/ScoreUtils.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/ScoreUtils.java:61]: This was missing all the time?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java:381]: Now wouldn't it be lovely if `AbstractScoreDirector` had a generic <Score_> type? :-) Inner and Abstract score directors, and their subclasses, are officially private code in 8, so adding more generics there does not affect users.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/InnerScoreDirector.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/InnerScoreDirector.java:81]: IMO `InnerScoreDirector` is really screaming for a `<Score_>` generic type.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/incremental/IncrementalScoreDirector.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/incremental/IncrementalScoreDirector.java:127]: The cast should not be necessary, as `ScoreDefinition` also has `<Score_>`.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/ConstraintSession.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/ConstraintSession.java:45]: Considering how often these two methods (constraint match total map, indictment map) repeat in various unrelated interfaces, perhaps these should be extracted into their own interface, which the other interfaces would then extend?

**Review by triceo — APPROVED** [review]
Love it!
Added some suggestions and observations.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java:686]: I fear this will go to far. That's a rabbit hole that will lead very deep.
Separate PR/experiment in any case. And let's first figure out if InnerScoreDirector should be score generified...

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/supply/SupplyManager.java
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/supply/SupplyManager.java:34]: 1) Yes, was a mistake
2) No: SupplyManager handles multiple Supple types.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/ScoreUtils.java
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/ScoreUtils.java:61]: Yes. Scary :)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java:381]: Yes, I am starting to think so too. Make InnerScoreDirector (not ScoreDirector!) Score generic. Separate PR in any case. Want to create a jira for it?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/incremental/IncrementalScoreDirector.java
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/incremental/IncrementalScoreDirector.java:127]: But as long as AbstractScoreDirector doesn't have a generic type, it ScoreDefinition isn't typed either. So can't fix in this PR.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/ConstraintSession.java
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/ConstraintSession.java:45]: I think this is premature optimizaton, that if we clean up some architecture, will become much less.

They repeat 3 times:
- ScoreExplenation - which is public API, so the extra interface inheritence will complicate it for users
- ScoreDirector
- ConstraintSession

I believe ConstraintSession is a needless indirection (that maybe even has a perf impact).
Same for ConstraintSessionFactory and ConstraintStreamScoreDirectorFactory.
We could maybe replace it by [Drools/Bavet]ConstraintStreamScoreDirector(Factory) extends AbstractConstraintStreamScoreDirector(Factory).
That experiment needs to be looked into first.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/director/AbstractScoreDirector.java:381]: https://issues.redhat.com/browse/PLANNER-2164

**Comment by ge0ffrey** [comment]
Feedback applied, waiting for CI now.

**Pull request merged**
